### PR TITLE
Release workspace

### PR DESCRIPTION
**Pull request by StephaneRob**
Add `release_workspace` role option. 

In build task, the release is copied from distillery release directory to `release_workspace`. (prevent download from remote to local)

In deploy task, the release is copied from `release_workspace` to `workspace`. (prevent upload from local to remote)

### CODE REVIEW

**Comment by brienw** [comment]
Thanks for the PR, @StephaneRob! I think this is a great solution. I wonder if it would be better if the config option were specified at the top level instead of per-role. For example:

```
use Bootleg.Config

config release_workspace: '/some/directory/'

role :build ...
role :app ...
```

having it as a config at that level would prevent there being two different settings that could possibly get out of sync. Like if a developer changes the setting for role `:build` but forgets to change it for role `:app`. 

If it was specified at the top level, I think we'd need a config keyword that makes it obvious that if set it will:
1) skip the local download
2) assume the same servers are configured for `build` and `app`
3) designate an interstitial directory for releases

While we noodle on that (feel free to give any ideas/suggestions you have as well) would you mind adding a test for this change?

**Comment by brienw** [comment]
Fixes [#177]

**Comment by StephaneRob** [comment]
@brienw My idea was to keep both roles independant to be able to allow remote `release_workspace` in the future and not limit the feature when build and app roles have the same server
ex:

```
role :app, release_wokspace: :build
```
could scp the release from the `release_workspace` on the `build` remote.

However, I agree with you the config at top level seems simpler but need to we assume that the same servers are configured for `build` and `app`.

I opened the PR to start the discussion, sure i will add tests for the solution you prefer.

**Comment by brienw** [comment]
@StephaneRob You bring up a good point about being able to use the release_workspace independent of whether another role uses it. And another benefit is it's much more straight forward to explain the option to users when used per-role.

so, 👍  from us.

Let us know if you need a tour of how the test env is set up. If you have Docker installed, the tests should set up all the necessary images automatically on first run - note that it can take a bit to build the initial test image. Subsequent runs will be much faster

File: test/bootleg/tasks/build_task_functional_test.exs [hunk 1]
```diff
@@ -38,6 +38,18 @@ defmodule Bootleg.Tasks.BuildTaskFunctionalTest do
...
   end
 
+  test "builds the application with a release_workspace path", %{hosts: [host], project_location: location} do
+    use Bootleg.Config
+
+    role :build, host.ip, release_workspace: "/home/#{host.user}/release_workspace", port: host.port
+
+    File.cd!(location, fn ->
+      capture_io(fn ->
+        invoke :build
```
Comment (brienw) [inline test/bootleg/tasks/build_task_functional_test.exs]: Thank you for the tests! I would feel more comfortable if there was an assertion here, maybe we can have an assertion that the tgz file exists in the remote workspace. - I also noticed that the test this one was based off of also doesn't actually have an assertion, i'll file an issue for that one

assertion can probably be something along the lines of:
```
...
invoke :build
assert [{:ok, ["#{Config.app()}.tar.gz"], 0, _}] = remote :build, "ls /home/#{host.user}/release_workspace"
...
```

File: test/bootleg/tasks/deploy_task_functional_test.exs [hunk 2]
```diff
@@ -40,4 +40,15 @@ defmodule Bootleg.Tasks.DeployTaskFunctionalTest do
...
     end)
   end
+
+  test "deploy/1 deploys the release to the target hosts from a remote release_workspace path", %{hosts: [host]} do
+    role :app, [host.ip], port: host.port, user: host.user, password: host.password,
+      silently_accept_hosts: true, workspace: "workspace", release_workspace: "/fixtures"
+
+    File.cd!("test/fixtures", fn ->
+      capture_io(fn ->
+        invoke :deploy
```
Comment (brienw) [inline test/bootleg/tasks/deploy_task_functional_test.exs]: same as above, please add an assertion to make sure that the deploy was successful. it doesn't have to be anything super complex, maybe just a quick check that one of the unpacked files exists where we would expect it to

**Comment by StephaneRob** [comment]
@brienw adding assertions on my code revealed an issue on current tests.

Actually setting role on `test` block doesn't work as expected, test still uses role configured on `setup`. So on `build_task_functional_test.exs` test, `invoke :build` does exactly the same than the first two tests. (same for the test with absolute workspace, it uses `"workspace"`).

I changed `Bootleg.FunctionalCase` to be able to specify a `@tag role_opts: %{}` on tests we want a specific option on role.

#### example for `build_task_functional_test.exs`:
```elixir
  # ...
  setup %{hosts: [host], role_opts: role_opts} do
    use Bootleg.Config
    config :app, :build_me
    config :version, "0.1.0"
    workspace = if role_opts[:workspace], do: role_opts[:workspace], else: "workspace"

    role :build, host.ip, port: host.port, user: host.user, password: host.password,
      silently_accept_hosts: true, workspace: workspace, identity: host.private_key_path, release_workspace: role_opts[:release_workspace]

    %{
      project_location: Fixtures.inflate_project
    }
  end

  @tag role_opts: %{release_workspace: "/home/me/release_workspace"}
  test "builds the application with a release_workspace path", %{hosts: [host], project_location: location} do
    # test
  end

  # ...
```

Let me know what do you think and if you want i push a commit with these changes

**Comment by brienw** [comment]
@StephaneRob Thank you for the updates! Sorry for the slow response, Rob and I are at Lonestar ElixirConf but will take a look when we get back next week!

**Comment by brienw** [comment]
I pulled this down earlier today and ran tests locally but they all blew up. i suspect that is probably related to the OSX update I did last night. I fiddle some bits and am running them again quick. PR looks good, so I'll likely merge this soon

**Comment by brienw** [comment]
Thank you for putting all the time into this @StephaneRob. We really appreciate it!

**Comment by StephaneRob** [comment]
Thanks for this great project!

**Pull request merged**